sds/src/proximity_keywords.rs

```diff
@@ -317,6 +346,101 @@ fn calculate_keyword_pattern(keyword: &str) -> Ast {
     })
 }
 
+#[derive(Clone, Copy)]
+enum CharType {
+    Regular,
+    Uppercase,
+    Separator,
+}
+
+fn get_chars_type(prev: &char, current: &char, next: &char) -> (CharType, CharType, CharType) {
+    let get_single_char_type = |c: &char| -> CharType {
+        let is_link_symbol = MULTI_WORD_KEYWORDS_LINK_CHARS.contains(c);
+        let is_uppercase_char = c.is_ascii_uppercase();
+
+        if is_link_symbol {
+            CharType::Separator
+        } else if is_uppercase_char {
+            CharType::Uppercase
+        } else {
+            CharType::Regular
+        }
+    };
+
+    (
+        get_single_char_type(prev),
+        get_single_char_type(current),
+        get_single_char_type(next),
+    )
+}
+
+fn standardize_path_chars<F>(chars: Vec<char>, mut push_character: F)
+where
+    F: FnMut(&char),
+{
+    let kw_length = chars.len();
+
+    if kw_length >= 1 {
+        push_character(&chars[0])
+    }
+
+    for chars in chars.windows(3) {
+        let current = &chars[1];
+        let (prev_char, current_char, next_char) = get_chars_type(&chars[0], &chars[1], &chars[2]);
+        match (prev_char, current_char, next_char) {
```

Comment:
It seems to me that `next_char` is not used, it could probably be dropped.

I did not try myself, but maybe you can use a simple for loop on `keyword.chars()` and keep track on the previous character type (if it exist). This way, the `char_list` won't need to be created in calculate_keyword_path_pattern, and possibly `kw_length >= 2` at the end can be dropped

Reply:
[Improve algorithm and reduce window to 2](https://github.com/DataDog/dd-sensitive-data-scanner/pull/52/commits/44c69740dc08a3c43ce532237f5ed8f402bc78b1)